### PR TITLE
Web assert add timeout

### PR DESCRIPTION
**Pull request by alexisLefebvre**
May be a solution for:

    https://github.com/minkphp/Mink/issues/606

I worked with @lenybernard in order to merge his changes to the current version of Mink, see the previous discussion for reference:

    https://github.com/Behat/MinkExtension/pull/210

(I put links in blockquote to avoid creating back links on these pages)

### CODE REVIEW

File: src/WebAssert.php [hunk 1]
```diff
@@ -46,12 +47,17 @@ public function __construct(Session $session)
...
+    public function addressEquals($page, $timeout = 10)
     {
-        $expected = $this->cleanUrl($page);
-        $actual = $this->getCurrentUrlPath();
+        $cleanedUrl = $this->cleanUrl($page);
+        $actual = null;
 
-        $this->assert($actual === $expected, sprintf('Current page is "%s", but "%s" expected.', $actual, $expected));
+        $this->assert(
+            $this->session->getPage()->waitFor($timeout, function () use ($cleanedUrl, &$actual) {
```
Comment (alexisLefebvre) [inline src/WebAssert.php]: @lenybernard Do you think that we should keep the `waitFor()` for the address? I'm pretty sure that the address of a page is not supposed to change once a page if loaded (if Mink wait for redirections, I didn't check that).

File: src/WebAssert.php [hunk 2]
```diff
@@ -61,12 +67,19 @@ public function addressEquals($page)
...
+        $cleanedUrl = $this->cleanUrl($page);
+        $actual = null;
 
-        $this->assert($actual !== $expected, sprintf('Current page is "%s", but should not be.', $actual));
+        $this->assert(
+            $this->session->getPage()->waitFor($timeout, function () use ($cleanedUrl, &$actual) {
+                $actual = $this->getCurrentUrlPath();
+                // TODO Don't fail on first try if address are different.
+                // TODO This condition is true and break the loop.
+                return $actual !== $cleanedUrl;
```
Comment (alexisLefebvre) [inline src/WebAssert.php]: @lenybernard I think that this condition can't work.

Let's say that the test expect the address to be different from `path/`:
- if the URL is `path/` during one second, then `/file`, the test will work (after trying during one second)
- if the URL is `/file` during one second, then `path/`, the test will work, the condition will be evaluated to true during the first loop iteration, it won't wait several seconds, so the test will be marked as passed incorrectly

File: src/WebAssert.php (hunk 1 above)
Comment (lenybernard) [reply src/WebAssert.php]: Actually, the address could change with javascript so I guess it can be useful

**Review by lenybernard — CHANGES_REQUESTED** [review]
Please run php-cs-fixer before

File: src/WebAssert.php [hunk 3]
```diff
@@ -28,6 +28,7 @@
 class WebAssert
 {
     protected $session;
+    const timeout = 5;
```
Comment (lenybernard) [inline src/WebAssert.php]: constant should be uppercased please

File: src/WebAssert.php [hunk 4]
```diff
@@ -61,12 +67,19 @@ public function addressEquals($page)
...
-        $expected = $this->cleanUrl($page);
-        $actual = $this->getCurrentUrlPath();
+        $cleanedUrl = $this->cleanUrl($page);
+        $actual = null;
 
-        $this->assert($actual !== $expected, sprintf('Current page is "%s", but should not be.', $actual));
+        $this->assert(
+            $this->session->getPage()->waitFor($timeout, function () use ($cleanedUrl, &$actual) {
+                $actual = $this->getCurrentUrlPath();
+                // TODO Don't fail on first try if address are different.
```
Comment (lenybernard) [inline src/WebAssert.php]: remove comments please

File: src/WebAssert.php [hunk 5]
```diff
@@ -378,22 +418,30 @@ public function responseNotMatches($regex)
...
+            $nodes = $givenNode->findAll($selectorType, $selector);
 
+            return intval($count) === count($nodes);
+        };
+        $result = $container->waitFor($timeout, $callback);
         $message = sprintf(
             '%d %s found on the page, but should be %d.',
             count($nodes),
-            $this->getMatchingElementRepresentation($selectorType, $selector, count($nodes) !== 1),
+            $this->getMatchingElementRepresentation($selectorType, $selector, count($nodes) > 1),
```
Comment (lenybernard) [inline src/WebAssert.php:442]: why ?

File: src/WebAssert.php [hunk 6]
```diff
@@ -402,17 +450,23 @@ public function elementsCount($selectorType, $selector, $count, ElementInterface
      * @param string           $selectorType element selector type (css, xpath)
      * @param string|array     $selector     element selector
      * @param ElementInterface $container    document to check against
+     * @param integer          $timeout      time to wait until fail
```
Comment (lenybernard) [inline src/WebAssert.php]: prefer float like waitFor method

File: tests/WebAssertTest.php [hunk 7]
```diff
@@ -5,6 +5,26 @@
 use Behat\Mink\Exception\ExpectationException;
 use Behat\Mink\WebAssert;
 
+/**
```
Comment (lenybernard) [inline tests/WebAssertTest.php]: don't add these comments

File: tests/WebAssertTest.php [hunk 8]
```diff
@@ -15,6 +35,7 @@ class WebAssertTest extends \PHPUnit_Framework_TestCase
      * @var WebAssert
      */
     private $assert;
+    const timeoutLimit = 0.11;
```
Comment (lenybernard) [inline tests/WebAssertTest.php]: explane here wtf is this value 🤓

Comment (lenybernard) [reply tests/WebAssertTest.php]: uppercase

File: src/WebAssert.php (hunk 3 above)
Comment (alexisLefebvre) [inline src/WebAssert.php]: This must be documented in CHANGES.md and in README.md.

**Review by lenybernard — CHANGES_REQUESTED** [review]
please add also some kind of docs in `CHANGES.md` and add a use case in the `README.md` to show off what we can do now with this amazing timeout and how we can configure it

**Pull request closed without merging**